core/corehttp/gateway_handler.go

```diff
@@ -227,92 +229,102 @@ func (i *gatewayHandler) getOrHeadHandler(ctx context.Context, w http.ResponseWr
 		return
 	}
 
-	links, err := i.api.Unixfs().Ls(ctx, resolvedPath)
+	nd, err := i.api.ResolveNode(ctx, resolvedPath)
 	if err != nil {
 		internalWebError(w, err)
 		return
 	}
 
-	// storage for directory listing
-	var dirListing []directoryItem
-	// loop through files
-	foundIndex := false
-	for _, link := range links {
-		if link.Name == "index.html" {
-			log.Debugf("found index.html link for %s", urlPath)
-			foundIndex = true
-
-			if urlPath[len(urlPath)-1] != '/' {
-				// See comment above where originalUrlPath is declared.
-				http.Redirect(w, r, originalUrlPath+"/", 302)
-				log.Debugf("redirect to %s", originalUrlPath+"/")
-				return
-			}
+	dirr, err := uio.NewDirectoryFromNode(i.node.DAG, nd)
+	if err != nil {
+		internalWebError(w, err)
+		return
+	}
 
-			dr, err := i.api.Unixfs().Cat(ctx, coreapi.ParseCid(link.Cid))
-			if err != nil {
-				internalWebError(w, err)
-				return
-			}
-			defer dr.Close()
+	ixnd, err := dirr.Find(ctx, "index.html")
+	switch {
+	case err == nil:
+		log.Debugf("found index.html link for %s", urlPath)
+
+		if urlPath[len(urlPath)-1] != '/' {
+			// See comment above where originalUrlPath is declared.
+			http.Redirect(w, r, originalUrlPath+"/", 302)
+			log.Debugf("redirect to %s", originalUrlPath+"/")
+			return
+		}
 
-			// write to request
-			http.ServeContent(w, r, "index.html", modtime, dr)
-			break
+		dr, err := i.api.Unixfs().Cat(ctx, coreapi.ParseCid(ixnd.Cid()))
+		if err != nil {
+			internalWebError(w, err)
+			return
 		}
+		defer dr.Close()
 
+		// write to request
+		http.ServeContent(w, r, "index.html", modtime, dr)
+		return
+	default:
+		internalWebError(w, err)
```

Comment:
In this case it isn't about keyword but its meaning. In this case `default` error is still an error, below it there is non-error handler.